cmake: FindZephyr-sdk: Find newest SDK version

nordicjm opened the pull request:
This improves the version checking for the zephyr SDK by searching for newer versions that the version that was detected. This works around the issue whereby the versions are contained in files with MD5 hashes which might be in any order.

Fixes #52234

nordicjm committed:
cmake: FindZephyr-sdk: Find newest SDK version

This improves the version checking for the zephyr SDK by searching
for newer versions that the version that was detected. This works
around the issue whereby the versions are contained in files with
MD5 hashes which might be in any order.

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

tejlmand reviewed (changes requested):
unfortunately this is not working in all cases:
```
$ tail -n +1 ~/.cmake/packages/Zephyr-sdk/*
==> /home/tora/.cmake/packages/Zephyr-sdk/8a6df3c3cd3af6a671c47f4639466a66 <==
/test/zephyr-sdk-0.16.6/cmake
==> /home/tora/.cmake/packages/Zephyr-sdk/d18bcc98b4b850ed9c8d40150420a8f0 <==
/test/zephyr-sdk-0.16.2/cmake
==> /home/tora/.cmake/packages/Zephyr-sdk/ed9f40aca023712f122056f8d8a8270b <==
/test/zephyr-sdk-0.15.3/cmake⏎                                     
```

results in `0.16.2` being picked up in my test:
```
rm -rf build; cmake -GNinja -DBOARD=nrf52840dk_nrf52840 -Bbuild samples/hello_world/ --trace-expand 2> cmake.trace
Running with expanded trace output on.
-- Application: /projects/github/ncs/zephyr/samples/hello_world
-- CMake version: 3.25.1
...
-- Found host-tools: zephyr 0.16.2 (/test/zephyr-sdk-0.16.2)
-- Found toolchain: zephyr 0.16.2 (/test/zephyr-sdk-0.16.2)
```

additional trace info:
```
3914 /test/zephyr-sdk-0.15.3/cmake/Zephyr-sdkConfig.cmake(10):  set(SDK_MAJOR_MINOR_MICRO 0.15.3 )       
  3915 /test/zephyr-sdk-0.15.3/cmake/Zephyr-sdkConfig.cmake(12):  get_filename_component(ZEPHYR_SDK_INSTALL_DIR /test/zephyr-sdk-0.15.3/cmake/.. ABSOLUTE )
  3916 /test/zephyr-sdk-0.15.3/cmake/Zephyr-sdkConfig.cmake(13):  set(ZEPHYR_SDK_INSTALL_DIR /test/zephyr-sdk-0.15.3 )
  3917 /test/zephyr-sdk-0.15.3/cmake/Zephyr-sdkConfig.cmake(14):  set(ZEPHYR_TOOLCHAIN_VARIANT zephyr )    
  3918 /test/zephyr-sdk-0.15.3/cmake/Zephyr-sdkConfig.cmake(17):  set(Zephyr-sdk_FOUND True )              
  3919 /test/zephyr-sdk-0.15.3/cmake/Zephyr-sdkConfig.cmake(18):  set(Zephyr-sdk_DIR /test/zephyr-sdk-0.15.3 )
  3920 /projects/github/ncs/zephyr/cmake/modules/FindZephyr-sdk.cmake(76):  set(NEWER_SDK_SEARCH 0 )       
  3921 /projects/github/ncs/zephyr/cmake/modules/FindZephyr-sdk.cmake(77):  while(0 EQUAL 0 )              
  3922 /projects/github/ncs/zephyr/cmake/modules/FindZephyr-sdk.cmake(78):  math(EXPR TMP_NEXT_VERSION_PATCH 3 + 1 )
  3923 /projects/github/ncs/zephyr/cmake/modules/FindZephyr-sdk.cmake(79):  set(Zephyr-sdk_FIND_VERSION 0.15.4 )
  3924 /projects/github/ncs/zephyr/cmake/modules/FindZephyr-sdk.cmake(82):  find_package(Zephyr-sdk 0.15.4 QUIET CONFIG PATHS /usr /usr/local /opt /home/tora /home/tora/.local /home/tora/.local/opt /home/tora/bin )
  3925 /test/zephyr-sdk-0.15.3/cmake/Zephyr-sdkConfigVersion.cmake(8):  file(READ /test/zephyr-sdk-0.15.3/cmake/../sdk_version SDK_VERSION )
  3926 /test/zephyr-sdk-0.15.3/cmake/Zephyr-sdkConfigVersion.cmake(9):  string(STRIP 0.15.3                
  3927  PACKAGE_VERSION )                                                                                  
  3928 /test/zephyr-sdk-0.15.3/cmake/Zephyr-sdkConfigVersion.cmake(11):  if(PACKAGE_VERSION VERSION_LESS PACKAGE_FIND_VERSION )
  3929 /test/zephyr-sdk-0.15.3/cmake/Zephyr-sdkConfigVersion.cmake(12):  set(PACKAGE_VERSION_COMPATIBLE FALSE )
  3930 /test/zephyr-sdk-0.16.2/cmake/Zephyr-sdkConfigVersion.cmake(8):  file(READ /test/zephyr-sdk-0.16.2/cmake/../sdk_version SDK_VERSION )
  3931 /test/zephyr-sdk-0.16.2/cmake/Zephyr-sdkConfigVersion.cmake(9):  string(STRIP 0.16.2                
  3932  PACKAGE_VERSION )                                                                                  
....
 3961 /test/zephyr-sdk-0.16.2/cmake/Zephyr-sdkConfig.cmake(18):  set(Zephyr-sdk_DIR /test/zephyr-sdk-0.16.2 )
  3962 /projects/github/ncs/zephyr/cmake/modules/FindZephyr-sdk.cmake(91):  if(2 LESS 4 )                  
  3963 /projects/github/ncs/zephyr/cmake/modules/FindZephyr-sdk.cmake(92):  set(NEWER_SDK_SEARCH 1 )       
  3964 /projects/github/ncs/zephyr/cmake/modules/FindZephyr-sdk.cmake(96):  set(NEWER_SDK_SEARCH )         
```
causing the search to abort even though `0.16.6` being available in the cache.
Note: local tests versions created just for testing the algorithm presented in this PR.

tejlmand commented:
also, what would happen in the event that the newer SDK version is *not* compatible with current Zephyr SDK ?

nordicjm commented:
I wast thinking about only finding the highest version with the patch e.g. if 0.15 is specified then the highest 0.15.x version, but alternatively we could specify a maximum supported version in a separate variable and update that for each release.

I think really this is a bit hackish and would be better for cmake to handle it but don't think that would be likely or if it was, to happen within a short timeframe.

tejlmand commented:
> I wast thinking about only finding the highest version with the patch e.g. if 0.15 is specified then the highest 0.15.x version

The I guess I did a good test, as it jumps to higher minor version.

> but alternatively we could specify a maximum supported version in a separate variable and update that for each release.

That would be `find_package()` implementation itself.
A package must report if it is not compatible with a required version, so that a `0.16.x` can report FALSE if it is incompatible when a `0.15.x` is requested. Not sure this is handled either in this PR 😉 

So if a package is backwards incompatible, then it should set `PACKAGE_VERSION_COMPATIBLE` to `False`.
Ref: https://cmake.org/cmake/help/latest/manual/cmake-packages.7.html#package-version-file

> I think really this is a bit hackish and would be better for cmake to handle it 

I completely agree. There are several use-cases we must ensure are working if we decide to make a fully working solution.
A trick we are using in `Zephyr` package is to look for version `0.0.0` with `EXACT`, cause that will always fail, and then we use the variable `<package>_CONSIDERED_CONFIGS` to select the Zephyr.

Maybe something similar wrt. Zephyr SDK can be done.
Get all versions by asking for `0.0.0`, and then work our way backwards from highest till lowest until there is a working match.
Ref:
https://github.com/zephyrproject-rtos/zephyr/blob/80f87b94806bef4f893b18f1dc8cea070382dd6b/share/zephyr-package/cmake/zephyr_package_search.cmake#L72
https://github.com/zephyrproject-rtos/zephyr/blob/80f87b94806bef4f893b18f1dc8cea070382dd6b/share/zephyr-package/cmake/zephyr_package_search.cmake#L82

nordicjm commented:
> Then I guess I did a good test, as it jumps to higher minor version.

I think you managed to stumble into a crafty bug! 0.15.2 searches for 0.15.3 or above and finds 0.16.0, which is actually a higher version, but it only checks the minor part - 3 vs 0... Should probably have used the version_compare one instead.

That 0.0.0 one looks like a good way to do it, I did read that I think it can return at most 4 versions in the list, need to test that and see if it is configurable

tejlmand commented:
> I think it can return at most 4 versions in the list

I have not noticed such limitation, but if you find such mentioning then please share.
But as it is a CMake list then I don't see why it should limit to 4.
That would be like CMake specifying that if it haven't found a matching package in the first 4 tries it then it will give up further tests.

nordicjm commented:
Have updated to search based on version 0.0.0

tejlmand commented on cmake/modules/FindZephyr-sdk.cmake:
```diff
@@ -60,7 +60,44 @@ if(("zephyr" STREQUAL ${ZEPHYR_TOOLCHAIN_VARIANT}) OR
+
+      include(${ZEPHYR_SDK_VERSION_CANDIDATE} NO_POLICY_SCOPE)
+
+      if("${PACKAGE_VERSION_COMPATIBLE}" STREQUAL "TRUE")
```
`PACKAGE_VERSION_COMPATIBLE` is a value which must be set by the package version file, as described:
> The version file must use these variables to check whether it is compatible or an exact match for the requested version and set the following variables with results:
> ...
> PACKAGE_VERSION_COMPATIBLE
>    True if version is compatible

so in principle this value value should be false, but because you manually loaded `Version` file and bypassed `find_package()` you have it available.
This makes this code harder to read and maintain.

Secondly, never do a CMake boolean test using a `STREQUAL`, cause the following values are also valid as `true`.
> True if the constant is 1, ON, YES, TRUE, Y, or a non-zero number (including floating point numbers)

nordicjm replied:
Fixed

tejlmand commented on cmake/modules/FindZephyr-sdk.cmake:
```diff
@@ -60,7 +60,44 @@ if(("zephyr" STREQUAL ${ZEPHYR_TOOLCHAIN_VARIANT}) OR
+    foreach(ZEPHYR_SDK_CANDIDATE ${Zephyr-sdk_CONSIDERED_CONFIGS})
+      string(REGEX REPLACE "\.cmake$" "Version.cmake" ZEPHYR_SDK_VERSION_CANDIDATE ${ZEPHYR_SDK_CANDIDATE})
+
+      include(${ZEPHYR_SDK_VERSION_CANDIDATE} NO_POLICY_SCOPE)
```
not fond of bypassing `find_package()` infrastructure this way when loading the Version file directly.

Such implementation should be used with great care.
Reason why it's used in https://github.com/zephyrproject-rtos/zephyr/blob/main/share/zephyr-package/cmake/zephyr_package_search.cmake is because that code handles multiple Zephyr within same workspace, and or Zephyrs which may not be registered in the CMake package registry.

The `FindZephyr-sdk.cmake` should only deal with Zephyr SDK in common search paths or those that are registered in the CMake package registry.

Why can't `find_package()` own infrastructure be used ?
For example by simply call `find_package(Zephyr-sdk <version-to-try>)` from highest version and down until a Zephyr SDK itself reports it is compatible.

nordicjm replied:
Fixed

tejlmand commented on cmake/modules/FindZephyr-sdk.cmake:
```diff
@@ -60,7 +60,44 @@ if(("zephyr" STREQUAL ${ZEPHYR_TOOLCHAIN_VARIANT}) OR
+                 $ENV{HOME}/.local/opt
+                 $ENV{HOME}/bin)
+
+    set(PACKAGE_FIND_VERSION ${Zephyr-sdk_FIND_VERSION})
```
This is something that `find_package()` itself takes care of, I really don't think we should start re-implementing a scaled down version of how `find_package()` works in this case.

nordicjm replied:
Fixed

tejlmand commented on cmake/modules/FindZephyr-sdk.cmake:
```diff
@@ -60,7 +60,44 @@ if(("zephyr" STREQUAL ${ZEPHYR_TOOLCHAIN_VARIANT}) OR
+
+    # Now that the best version has been found, find the exact Zephyr SDK of this
+    # version and load it.
+    find_package(Zephyr-sdk ${COMPATIBLE_SDK_VERSION} EXACT REQUIRED QUIET CONFIG PATHS
```
why not simply order the `Zephyr-sdk_CONSIDERED_CONFIGS` list from newest to oldest, and then call this `find_package(<version-to-try> EXACT)` with newest first, until `PACKAGE_VERSION_COMPATIBLE` is true ?

nordicjm replied:
Fixed

tejlmand reviewed (changes requested):
General comment, we should ensure that we don't re-run this lookup on incremental builds, but that the Zephyr SDK version is properly cached and re-used between CMake invocation.

I haven't verified if this is the case, but just mentioning this to ensure it's properly tested in various setups.

nordicjm commented:
If application is configured, then reconfigured e.g. with `cmake .` then the zephyr-sdk find function is not called again (i.e. cached value is used)

nordicjm commented:
@stephanosio is this something we want in for 3.3 so that if someone has e.g. sdk-ng 0.15.0 and 0.15.2 it will use 0.15.2 instead of 0.15.0, or should it wait for 3.4?

stephanosio commented:
This is not fixing a bug. It will need to wait for 3.4.

nordicjm commented:
@stephanosio can we revisit this? This is fixing a bug for me at least. I have sdk-ng 0.15.0 with just the arm tools, and 0.15.2 with all the tools, whenever I configure anything that is not for the arm platform, it fails with:
```
CMake Error at /tmp/aa/zephyr/cmake/compiler/gcc/target.cmake:12 (message):
  C compiler
  /opt/zephyr-sdk-0.15.0/riscv64-zephyr-elf/bin/riscv64-zephyr-elf-gcc not
  found - Please check your toolchain installation
```
Which kind of is a bug

stephanosio commented:
> @stephanosio can we revisit this? This is fixing a bug for me at least. I have sdk-ng 0.15.0 with just the arm tools, and 0.15.2 with all the tools, whenever I configure anything that is not for the arm platform, it fails with:
> 
> ```
> CMake Error at /tmp/aa/zephyr/cmake/compiler/gcc/target.cmake:12 (message):
>   C compiler
>   /opt/zephyr-sdk-0.15.0/riscv64-zephyr-elf/bin/riscv64-zephyr-elf-gcc not
>   found - Please check your toolchain installation
> ```
> 
> Which kind of is a bug

I leave the decision on whether that should be considered a bug up to @tejlmand since he is the maintainer for the build system and, IIRC, he had a different view on this matter. If he considers this a bug, I have no objection to merging this.

stephanosio commented on cmake/modules/FindZephyr-sdk.cmake:
```diff
@@ -60,14 +60,50 @@ if(("zephyr" STREQUAL ${ZEPHYR_TOOLCHAIN_VARIANT}) OR
                  $ENV{HOME}/.local/opt
                  $ENV{HOME}/bin)
+
+    # Remove duplicat entries and sort naturally in descending order.
```
typo

```suggestion
    # Remove duplicate entries and sort naturally in descending order.
```

nordicjm replied:
Fixed

stephanosio commented on cmake/modules/FindZephyr-sdk.cmake:
```diff
@@ -60,14 +60,50 @@ if(("zephyr" STREQUAL ${ZEPHYR_TOOLCHAIN_VARIANT}) OR
+      # This means no compatible Zephyr SDK versions were found, set the version
+      # back to the minimum version so that it is displayed in the error text.
+      find_package(Zephyr-sdk ${Zephyr-sdk_FIND_VERSION} EXACT REQUIRED QUIET CONFIG PATHS
+                   /usr
+                   /usr/local
+                   /opt
+                   $ENV{HOME}
+                   $ENV{HOME}/.local
+                   $ENV{HOME}/.local/opt
+                   $ENV{HOME}/bin)
```
These default search paths are repeated three times in this file. It would be nice if this could be put into a variable (if possible).

nordicjm replied:
Fixed

tejlmand reviewed (changes requested):
Has this been tested for situations where a given Zephyr SDK version is marked as not compatible to the requested one ?
Cause to me, this code seems to always just pick latest Zephyr SDK, compatible or not.

nordicjm commented:
> Has this been tested for situations where a given Zephyr SDK version is marked as not compatible to the requested one ?

There is no such system at present, this is the code from the 0.15.2 sdk-ng cmake file:
```
file(READ ${CMAKE_CURRENT_LIST_DIR}/../sdk_version SDK_VERSION)
string(STRIP ${SDK_VERSION} PACKAGE_VERSION)

if(PACKAGE_VERSION VERSION_LESS PACKAGE_FIND_VERSION)
  set(PACKAGE_VERSION_COMPATIBLE FALSE)
else()
  # Currently, this Zephyr SDK is expected to work with any Zephyr project
  # requiring this version or any older version.
  #
  # In case this version is no longer backward compatible then this is the place
  # to test, for example as
  # set(ZEPHYR_MINIMUM_COMPATIBLE_VERSION 0.11)
  # if(PACKAGE_FIND_VERSION VERSION_LESS ZEPHYR_MINIMUM_COMPATIBLE_VERSION)
  #   set(PACKAGE_VERSION_COMPATIBLE FALSE)
  #   return()
  # endif()

  set(PACKAGE_VERSION_COMPATIBLE TRUE)
  if(PACKAGE_FIND_VERSION STREQUAL PACKAGE_VERSION)
    set(PACKAGE_VERSION_EXACT TRUE)
  endif()
endif()
```

tejlmand commented:
> There is no such system at present

Of course not, just hack together an unofficial Zephyr SDK where you take a real Zephyr SDK and manipulates it's version file and even mark it as incompatible.

That's what I have done when testing your PR, no need to wait till an official Zephyr SDK release for testing code flow 😉

nordicjm commented:
> Has this been tested for situations where a given Zephyr SDK version is marked as not compatible to the requested one ?
> Cause to me, this code seems to always just pick latest Zephyr SDK, compatible or not.

Tested, works fine.

Normal:
```
Loading Zephyr default modules (Zephyr repository).
-- Application: /tmp/aa/zephyr/samples/hello_world
-- CMake version: 3.25.2
-- Found Python3: /usr/bin/python3.10 (found suitable exact version "3.10.9") found components: Interpreter 
-- Cache files will be written to: /home/jamie/.cache/zephyr
-- Zephyr version: 3.2.99 (/tmp/aa/zephyr)
-- Found west (found suitable version "0.14.0", minimum required is "0.7.1")
-- Board: nrf52840dk_nrf52840
-- ZEPHYR_TOOLCHAIN_VARIANT not set, trying to locate Zephyr SDK
-- Found host-tools: zephyr 0.15.2 (/opt/zephyr-sdk-0.15.2)
-- Found toolchain: zephyr 0.15.2 (/opt/zephyr-sdk-0.15.2)
-- Found Dtc: /opt/zephyr-sdk-0.15.2/sysroots/x86_64-pokysdk-linux/usr/bin/dtc (found suitable version "1.6.0", minimum required is "1.4.6") 
-- Found BOARD.dts: /tmp/aa/zephyr/boards/arm/nrf52840dk_nrf52840/nrf52840dk_nrf52840.dts
```

With following change to `/opt/zephyr-sdk-0.15.2/cmake/Zephyr-sdkConfigVersion.cmake`:
```
   set(ZEPHYR_MINIMUM_COMPATIBLE_VERSION 4.11)
   if(PACKAGE_FIND_VERSION VERSION_LESS ZEPHYR_MINIMUM_COMPATIBLE_VERSION)
     set(PACKAGE_VERSION_COMPATIBLE FALSE)
     return()
   endif()
```
gives:
```
Loading Zephyr default modules (Zephyr repository).
-- Application: /tmp/aa/zephyr/samples/hello_world
-- CMake version: 3.25.2
-- Found Python3: /usr/bin/python3.10 (found suitable exact version "3.10.9") found components: Interpreter 
-- Cache files will be written to: /home/jamie/.cache/zephyr
-- Zephyr version: 3.2.99 (/tmp/aa/zephyr)
-- Found west (found suitable version "0.14.0", minimum required is "0.7.1")
-- Board: nrf52840dk_nrf52840
-- ZEPHYR_TOOLCHAIN_VARIANT not set, trying to locate Zephyr SDK
-- Found host-tools: zephyr 0.15.0 (/opt/zephyr-sdk-0.15.0)
-- Found toolchain: zephyr 0.15.0 (/opt/zephyr-sdk-0.15.0)
-- Found Dtc: /opt/zephyr-sdk-0.15.0/sysroots/x86_64-pokysdk-linux/usr/bin/dtc (found suitable version "1.6.0", minimum required is "1.4.6") 
-- Found BOARD.dts: /tmp/aa/zephyr/boards/arm/nrf52840dk_nrf52840/nrf52840dk_nrf52840.dts
-- Generated zephyr.dts: /tmp/aa/zephyr/samples/hello_world/_AA/zephyr/zephyr.dts
```

nordicjm commented on cmake/modules/FindZephyr-sdk.cmake:
```diff
@@ -60,14 +60,39 @@ if(("zephyr" STREQUAL ${ZEPHYR_TOOLCHAIN_VARIANT}) OR
+      find_package(Zephyr-sdk ${ZEPHYR_SDK_CANDIDATE} EXACT QUIET CONFIG PATHS ${ZEPHYR_SDK_SEARCH_PATHS})
+
+      if ("${Zephyr-sdk_FOUND}" EQUAL "1")
+        break()
```
Can't do a return because this is not a function. And in actuality if you do a return here, it then breaks the whole build because it no longer caches the zephyr SDK directory:
```
Kconfig header saved to '/tmp/aa/zephyr/samples/hello_world/_AA/zephyr/include/generated/autoconf.h'
-- Configuring done
You have changed variables that require your cache to be deleted.
Configure will be re-run and you may have to reset some variables.
The following variables have changed:
CMAKE_C_COMPILER= /opt/zephyr-sdk-0.15.0/arm-zephyr-eabi/bin/arm-zephyr-eabi-gcc
CMAKE_C_COMPILER= /opt/zephyr-sdk-0.15.0/arm-zephyr-eabi/bin/arm-zephyr-eabi-gcc
CMAKE_C_COMPILER= /opt/zephyr-sdk-0.15.0/arm-zephyr-eabi/bin/arm-zephyr-eabi-gcc

Loading Zephyr default modules (Zephyr repository).
-- Application: /tmp/aa/zephyr/samples/hello_world
-- CMake version: 3.25.2
-- Found Python3: /usr/bin/python3.10 (found suitable exact version "3.10.9") found components: Interpreter 
-- Cache files will be written to: /home/jamie/.cache/zephyr
-- Zephyr version: 3.2.99 (/tmp/aa/zephyr)
-- Found west (found suitable version "0.14.0", minimum required is "0.7.1")
CMake Error at /tmp/aa/zephyr/cmake/modules/extensions.cmake:2520 (message):
  BOARD is not being defined on the CMake command-line, in the environment or
  by the app.
Call Stack (most recent call first):
  /tmp/aa/zephyr/cmake/modules/boards.cmake:52 (zephyr_check_cache)
  /tmp/aa/zephyr/cmake/modules/zephyr_default.cmake:113 (include)
  /tmp/aa/zephyr/share/zephyr-package/cmake/ZephyrConfig.cmake:66 (include)
  /tmp/aa/zephyr/share/zephyr-package/cmake/ZephyrConfig.cmake:117 (include_boilerplate)
  /home/jamie/sysbuild/zephyr/share/zephyr-package/cmake/zephyr_package_search.cmake:117 (include)
  /home/jamie/sysbuild/zephyr/share/zephyr-package/cmake/ZephyrConfig.cmake:152 (check_zephyr_package)
  CMakeLists.txt:5 (find_package)


-- Configuring incomplete, errors occurred!
FAILED: build.ninja 
/usr/bin/cmake --regenerate-during-build -S/tmp/aa/zephyr/samples/hello_world -B/tmp/aa/zephyr/samples/hello_world/_AA
ninja: error: rebuilding 'build.ninja': subcommand failed
｢22 /tmp/aa/zephyr/samples/hello_world/_AA (sdkverfix)｣$ ninja
ninja: error: build.ninja:30: loading 'CMakeFiles/rules.ninja': No such file or directory
include CMakeFiles/rules.ninja
                              ^ near here

```

nordicjm commented on cmake/modules/FindZephyr-sdk.cmake:
```diff
@@ -60,14 +60,39 @@ if(("zephyr" STREQUAL ${ZEPHYR_TOOLCHAIN_VARIANT}) OR
+    if (NOT ("${Zephyr-sdk_FOUND}" EQUAL "1"))
+      # This means no compatible Zephyr SDK versions were found, set the version
+      # back to the minimum version so that it is displayed in the error text.
+      find_package(Zephyr-sdk ${Zephyr-sdk_FIND_VERSION} EXACT REQUIRED QUIET CONFIG PATHS ${ZEPHYR_SDK_SEARCH_PATHS})
```
Fixed

nordicjm committed:
doc: release: 3.4: Add note on fixed Zephyr SDK toolchain issue

Adds a note on a fixed issue whereby the latest Zephyr SDK
toolchain was not used depending upon hash values in the cmake
registry.

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

kartben commented on doc/releases/release-notes-3.4.rst:
```diff
@@ -115,6 +115,9 @@ Boards & SoC Support
 Build system and infrastructure
 *******************************
 
+* An issue whereby older versions of the Zephyr SDK toolchain were used
```
For consistency with the way release note items are typically worded, please change to:

Fixed an issue whereby [...] instead of the latest

nordicjm replied:
Fixed

tejlmand commented on cmake/modules/FindZephyr-sdk.cmake:
```diff
@@ -60,14 +60,41 @@ if(("zephyr" STREQUAL ${ZEPHYR_TOOLCHAIN_VARIANT}) OR
+
+    # Remove duplicate entries and sort naturally in descending order.
+    list(REMOVE_DUPLICATES Zephyr-sdk_CONSIDERED_VERSIONS)
+    list(SORT Zephyr-sdk_CONSIDERED_VERSIONS COMPARE NATURAL ORDER DESCENDING)
+
+    # Loop over each found Zepher SDK version until one is found that is compatible.
+    foreach(ZEPHYR_SDK_CANDIDATE ${Zephyr-sdk_CONSIDERED_VERSIONS})
+      if("${ZEPHYR_SDK_CANDIDATE}" VERSION_GREATER_EQUAL "${Zephyr-sdk_FIND_VERSION}")
+        find_package(Zephyr-sdk ${ZEPHYR_SDK_CANDIDATE} EXACT QUIET CONFIG PATHS ${ZEPHYR_SDK_SEARCH_PATHS})
```
this will cause you to loop from newest to oldest, like `3.0;2.0;1.0` so on first loop you invoke:
```
find_package(Zephyr-sdk 3.0 EXACT ...)
```
but of course the 3.0 installation of Zephyr SDK is compatible with requested version 3.0, so that shall always return true.

What you should do is to test each candidate against the `Zephyr-sdk_FIND_VERSION`, and in order to do so, you must use `Zephyr-sdk_CONSIDERED_CONFIGS` list, like this:
```
  foreach(zephyr_sdk_config ${Zephyr-sdk_CONSIDERED_CONFIGS})
    <adjust path of zephyr_sdk_config, to let's say zephyr_sdk_config_path>
    find_package(Zephyr-sdk ${Zephyr-sdk_FIND_VERSION} EXACT QUIET CONFIG PATHS <zephyr_sdk_config_path> NO_DEFAULT_PATH)
```
that way you'll test each candidate individually to see if it is compatible with the build system requested `Zephyr-sdk_FIND_VERSION`.

For inspiration on the `CONSIDERED_CONFIGS` path handling, find inspiration here:
https://github.com/zephyrproject-rtos/zephyr/blob/ca3a4f5f5b357c843d7f07bef443c6ca1945c249/share/zephyr-package/cmake/zephyr_package_search.cmake#L82-L85
Note: that code cannot be used as-is, but serve as inspiration for: `<adjust path of zephyr_sdk_config, ...>`

nordicjm replied:
Fixed

tejlmand commented on cmake/modules/FindZephyr-sdk.cmake:
```diff
@@ -60,20 +60,65 @@ if(("zephyr" STREQUAL ${ZEPHYR_TOOLCHAIN_VARIANT}) OR
   SET(CMAKE_FIND_PACKAGE_SORT_ORDER ${CMAKE_FIND_PACKAGE_SORT_ORDER_CURRENT})
 endif()
 
+# Clean up temp variables
+set(ZEPHYR_SDK_FOUND_VERSIONS)
+set(ZEPHYR_SDK_FOUND_CONFIGS)
+set(ZEPHYR_SDK_CURRENT_INDEX)
+set(ZEPHYR_SDK_CURRENT_CHECK_PATH)
```
we generally use lower case for internal vars to make it easier to distinguish those from global variables.

nordicjm replied:
Fixed

tejlmand reviewed (dismissed):
lgtm, but would be nice if you would find time to use lower case for internally used variables.

evgeniy-paltsev commented:
Just to be sure - do I understand right that we don't change the behavior if we select SDK via `ZEPHYR_SDK_INSTALL_DIR`?
 So the exact version which is specified in `ZEPHYR_SDK_INSTALL_DIR` will be used?

tejlmand commented on cmake/modules/FindZephyr-sdk.cmake:
```diff
@@ -60,20 +60,66 @@ if(("zephyr" STREQUAL ${ZEPHYR_TOOLCHAIN_VARIANT}) OR
+    list(SORT Zephyr-sdk_CONSIDERED_VERSIONS COMPARE NATURAL ORDER DESCENDING)
+
+    # Loop over each found Zepher SDK version until one is found that is compatible.
+    foreach(zephyr_sdk_canidate ${Zephyr-sdk_CONSIDERED_VERSIONS})
```
canidate ?
```suggestion
    foreach(zephyr_sdk_candidate ${Zephyr-sdk_CONSIDERED_VERSIONS})
```
and other locations as well.

nordicjm replied:
Fixed

tejlmand commented:
>  So the exact version which is specified in ZEPHYR_SDK_INSTALL_DIR will be used?

Correct, ref: https://github.com/nordicjm/zephyr/blob/96182d1ff55e3766eb72b9fc3322414005a6f701/cmake/modules/FindZephyr-sdk.cmake#L51-L58
